odoo/tools/image.py

```diff
@@ -26,132 +27,113 @@
-    if image.mode not in ["1", "L", "P", "RGB", "RGBA"] or (filetype == 'JPEG' and image.mode == 'RGBA'):
-        image = image.convert("RGB")
+    image = base64_to_image(base64_source)
 
-    background_stream = io.BytesIO()
-    image.save(background_stream, filetype)
-    return codecs.encode(background_stream.getvalue(), encoding)
+    w, h = image.size
+    if verify_resolution and w * h > IMAGE_MAX_RESOLUTION:
+        raise ValueError(_("Image size excessive, uploaded images must be smaller than %s million pixels.") % str(IMAGE_MAX_RESOLUTION / 10e6))
 
+    # get the format of the original image (must be done before resize)
+    format = (format or image.format).upper()
+    if format == 'BMP':
+        format = 'PNG'
+    elif format not in ['PNG', 'JPEG', 'GIF']:
+        format = 'JPEG'
 
-def image_resize_and_sharpen(image, size, preserve_aspect_ratio=False, factor=2.0, upper_limit=False):
-    """
-        Create a thumbnail by resizing while keeping ratio.
-        A sharpen filter is applied for a better looking result.
+    opt = {'format': format}
 
-        :param image: PIL.Image.Image()
-        :param size: 2-tuple(width, height)
-        :param preserve_aspect_ratio: boolean (default: False)
-        :param factor: Sharpen factor (default: 2.0)
-    """
-    origin_mode = image.mode
-    if image.mode != 'RGBA':
-        image = image.convert('RGBA')
-    image.thumbnail(size, Image.ANTIALIAS)
-    if preserve_aspect_ratio:
-        size = image.size
-    sharpener = ImageEnhance.Sharpness(image)
-    resized_image = sharpener.enhance(factor)
-    # create a transparent image for background and paste the image on it
-    if upper_limit:
-        image = Image.new('RGBA', (size[0], size[1]-3), (255, 255, 255, 0)) # FIXME temporary fix for trimming the ghost border.
-    else:
-        image = Image.new('RGBA', size, (255, 255, 255, 0))
-    image.paste(resized_image, ((size[0] - resized_image.size[0]) // 2, (size[1] - resized_image.size[1]) // 2))
-
-    if image.mode != origin_mode:
-        image = image.convert(origin_mode)
-    return image
-
-
-def image_save_for_web(image, fp=None, format=None):
-    """
-        Save image optimized for web usage.
+    if size and (size[0] or size[1]):
+        w, h = image.size
+        asked_width = size[0] or (w * size[1]) // h
+        asked_height = size[1] or (h * size[0]) // w
 
-        :param image: PIL.Image.Image()
-        :param fp: File name or file object. If not specified, a bytestring is returned.
-        :param format: File format if could not be deduced from image.
-    """
-    opt = dict(format=image.format or format)
-    if image.format == 'PNG':
-        opt.update(optimize=True)
+        if crop:
+            new_w, new_h = asked_width, asked_height
+
+            if new_w > w:
+                new_w, new_h = w, (new_h * w) // new_w
+            if new_h > h:
+                new_w, new_h = (new_w * h) // new_h, h
+
+            if type == 'top':
```

Comment:
always false. `type` is the builtin function https://docs.python.org/3.5/library/functions.html#type

Reply:
Good catch, the code was coming from another function where type was shadowed by a local variable.